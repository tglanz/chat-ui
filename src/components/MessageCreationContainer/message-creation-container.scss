.message-creation-container {
    /* pin */
    position: absolute;
    bottom: 3%;
    right: 1%;

    /* shadow */
    -webkit-box-shadow: 6px 7px 6px 0px rgba(0,0,0,0.75);
    -moz-box-shadow: 6px 7px 6px 0px rgba(0,0,0,0.75);
    box-shadow: 6px 7px 6px 0px rgba(0,0,0,0.75);

    /* round a bit */
    -webkit-border-radius: 4px;
    -moz-border-radius: 4px;
    border-radius: 4px;

    width: 30%;
    padding: 4px 2px 0 2px;    
}