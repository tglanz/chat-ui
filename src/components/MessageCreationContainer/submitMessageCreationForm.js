const validation = ({
    username,
    text, setText,
    setErrorMessage
}) => {
    if (!username){
        setErrorMessage("Missing username");
        return false;
    }
    
    if (!text){
        setErrorMessage("Missing text");
        return false;
    }

    return true;
}

// Happens upon submission
export default ({
    username,
    text, setText,

    setErrorMessage, emitMessage,

    randomlyProvideAvatar,
    setAvatar, usersAvatars, setUsersAvatars
}) => {

    if (!validation({ username, text, setText, setErrorMessage })){
        return;
    }

    let avatar = usersAvatars[username] || randomlyProvideAvatar();

    if (!usersAvatars.hasOwnProperty(username)){
        setUsersAvatars({
            ...usersAvatars,
            [username]: avatar
        })
    }

    setAvatar(avatar);
    emitMessage({ username, text, avatar, timestamp: new Date().getTime() });

    setErrorMessage("");
    setText("");
}