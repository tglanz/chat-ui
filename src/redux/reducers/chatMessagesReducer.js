import { Types } from '../actions/chatMessagesActions';

const initialState = [];

export default (state = initialState, action) => {
    switch (action.type){
        case Types.RECEIVED:
            return state.concat(action.payload.message);
        default:
            return state;
    }
}