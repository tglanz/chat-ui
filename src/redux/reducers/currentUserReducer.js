import { Types } from '../actions/currentUserActions';

const initialState = {
    username: ''
};

export default (state = initialState, action) => {
    switch (action.type){
        case Types.USERNAME_CHANGED:
            return {...state, username: action.payload.username };
        default:
            return state;
    }
}