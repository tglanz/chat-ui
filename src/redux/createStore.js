import { createStore } from 'redux';
import reducers from './reducers';
    
export default () => {
    // middlewares here
    return createStore(reducers);
}