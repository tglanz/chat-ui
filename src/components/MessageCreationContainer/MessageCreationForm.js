
import React from 'react'
import { connect } from 'react-redux';
import { compose, withState, withHandlers } from 'recompose';
import { Form, Image, Icon, Header, Message } from 'semantic-ui-react'

import withPersistentState from '../hoc/withPersistentState';
import withRandomAvatarProvider from '../hoc/withRandomAvatarProvider';

import * as currentUserActions from '../../redux/actions/currentUserActions';

import './message-creation-container.scss';
import submitMessageCreationForm from './submitMessageCreationForm';
import { bindActionCreators } from 'redux';

const MessageCreationForm = ({
    emitMessage,
    onSubmit,
    errorMessage, setErrorMessage,    
    username, setUsername, onUsernameChanged,
    text, setText,
    avatar, setAvatar
}) => <Form
    error={ errorMessage !== "" }
    onSubmit={ onSubmit }>

    <Header as='h2'>
        Send A Message
        <Image circular floated={'right'} src={ avatar } />
    </Header>

    <Form.Input
        name="username"
        label="Your username"
        icon={ <Icon name="user" /> }
        placeholder="Your username" type="text"
        value={username}
        onChange={ (e, { value }) => {
            setUsername(value);
            // onUsernameChanged(value); // DONT wanna put it here
        } }/>

    <Form.TextArea
        name="text"
        value={text}
        onChange={ (e, { value }) => setText(value) }/>

    <Message
      error
      header='Oops'
      content={ errorMessage }
    />

    <Form.Button icon labelPosition='left'>Send
        <Icon name='send' />
    </Form.Button>
</Form>

// Switch with redux connect any time :) no component state, just props
export default compose(
    withRandomAvatarProvider(),
    withPersistentState('username', 'setUsername', '', 'chat-ui/username'),
    withPersistentState('usersAvatars', 'setUsersAvatars', '', 'chat-ui/users-avatars'),
    withState('avatar', 'setAvatar', ''),
    withState('text', 'setText', ''),
    withState('errorMessage', 'setErrorMessage', ''),
    connect(
        state => {},
        dispatch => ({
            onUsernameChanged: bindActionCreators(currentUserActions.usernameChanged, dispatch)
        })
    ),
    withHandlers({
        onSubmit: props => event => {
            props.onUsernameChanged(props.username);
            submitMessageCreationForm(props)
        }
    })
)(MessageCreationForm);