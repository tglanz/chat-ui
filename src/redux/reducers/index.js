import { combineReducers } from 'redux';

import chatMessages from './chatMessagesReducer';
import currentUser from './currentUserReducer';

export default combineReducers({
    chatMessages,
    currentUser
});