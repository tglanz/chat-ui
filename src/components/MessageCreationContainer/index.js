
import React from 'react'

import './message-creation-container.scss';

import MessageCreationForm from './MessageCreationForm';

export default () => <div className={"message-creation-container"}>
    <MessageCreationForm />
</div>