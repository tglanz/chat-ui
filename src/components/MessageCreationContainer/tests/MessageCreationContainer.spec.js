import React from 'react';
import renderer from 'react-test-renderer';

import { createStore } from 'redux';
import { Provider } from 'react-redux';
import reducers from '../../../redux/reducers';

import MessageCreationContainer from '../';

describe("Container", () => {
    it("render", () => {
        const tree = renderer
            .create(<Provider store={ createStore(reducers) }>
                <MessageCreationContainer />
            </Provider>)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });
})