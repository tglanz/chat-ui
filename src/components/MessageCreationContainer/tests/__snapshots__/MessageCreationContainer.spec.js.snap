// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container render 1`] = `
<div
  className="message-creation-container"
>
  <form
    action={undefined}
    className="ui form"
    onSubmit={[Function]}
  >
    <h2
      className="ui header"
    >
      Send A Message
      <img
        alt={undefined}
        className="ui circular right floated image"
        height={undefined}
        src=""
        width={undefined}
      />
    </h2>
    <div
      className="field"
    >
      <label
        htmlFor={undefined}
      >
        Your username
      </label>
      <div
        className="ui icon input"
        content={undefined}
      >
        <input
          disabled={undefined}
          name="username"
          onChange={[Function]}
          placeholder="Your username"
          required={undefined}
          tabIndex={undefined}
          type="text"
          value=""
        />
        <i
          aria-hidden="true"
          className="user icon"
        />
      </div>
    </div>
    <div
      className="field"
    >
      <textarea
        content={undefined}
        disabled={undefined}
        name="text"
        onChange={[Function]}
        onInput={[Function]}
        required={undefined}
        rows={3}
        style={
          Object {
            "resize": "",
          }
        }
        type={undefined}
        value=""
      />
    </div>
    <div
      className="ui error message"
    >
      <div
        className="content"
      >
        <div
          className="header"
        >
          Oops
        </div>
        <p>
          
        </p>
      </div>
    </div>
    <div
      className="field"
    >
      <button
        className="ui icon left labeled button"
        disabled={undefined}
        onClick={[Function]}
        required={undefined}
        role="button"
        tabIndex={undefined}
        type={undefined}
      >
        Send
        <i
          aria-hidden="true"
          className="send icon"
        />
      </button>
    </div>
  </form>
</div>
`;
